lib/gitsh/tab_completion/dsl/option_transition_factory.rb

```diff
@@ -0,0 +1,60 @@
+require 'gitsh/tab_completion/automaton'
+require 'gitsh/tab_completion/matchers/unknown_option_matcher'
+
+module Gitsh
+  module TabCompletion
+    module DSL
+      class OptionTransitionFactory
+        def build(start_state, options = {})
+          Invocation.new(start_state, options).call
```

Comment:
I think the thing I like about that pattern is the clear separation between setting up an object, and doing work. We'd lose that separation here, but the result is overall less complex than the nested class, so I'll go with it.